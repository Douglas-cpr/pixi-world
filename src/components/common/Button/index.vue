<script setup lang="ts">

defineProps({
  onClick: {
    type: Function,
    required: true
  },
  disabled: {
    type: Boolean,
    default: false
  },
  type: {
    type: String,
    default: 'primary'
  },
  height: {
    type: Number,
    default: 35
  },
  width: {
    type: Number,
    default: 100
  }
})
</script>


<template>
<button
  @click="() => onClick()"
  :disabled="disabled"
  class="common__button"
  :class="{
    'common__button--primary': type === 'primary',
    'common__button--secondary': type === 'secondary',
  }"

  :style="{
    height: height + 'px',
    width: width + 'px'
  }">
  <slot></slot>
</button>
</template>

<style scoped>
.common__button {
  cursor: pointer;
  border: none;
  border-radius: 4rem;
  font-weight: 600;
}

.common__button--primary {
  color: #8C72FF;
  background-color: #fff;
}

.common__button--secondary {
  color: #fff;
  background-color: #8C72FF;
}
</style>