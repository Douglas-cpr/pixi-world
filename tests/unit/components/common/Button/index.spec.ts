import Button from '@/components/factories/Button/index.vue'
import { shallowMount, VueWrapper } from '@vue/test-utils'

type ButtonProps = {
  onClick: () => void
  disabled?: boolean
  type?: 'primary' | 'secondary'
  height?: number
  width?: number
}

describe('<Button />', () => {
  let wrapper: VueWrapper<any>;

  const createComponent = (buttonProps: ButtonProps) => {
    wrapper = shallowMount(Button, {
      props: {
        ...buttonProps
      }
    })
  }

  const createComponentWithDefaultProps = () => {
    createComponent({
      onClick: () => {},
      disabled: false,
      type: 'primary',
      height: 40,
      width: 100
    })
  }

  afterEach(() => {
    wrapper.unmount()
  })

  it('should render correctly', () => {
    createComponentWithDefaultProps()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should call onClick prop when clicked', () => {
    const onClick = jest.fn()
    createComponent({
      onClick,
    })

    wrapper.find('button').trigger('click')

    expect(onClick).toHaveBeenCalled()
  })

  it('should have a primary class when type is primary', () => {
    createComponent({
      onClick: () => {},
      type: 'primary'
    })

    expect(wrapper.classes()).toContain('common__button--primary')
  })

  it('should have a secondary class when type is secondary', () => {
    createComponent({
      onClick: () => {},
      type: 'secondary'
    })

    expect(wrapper.classes()).toContain('common__button--secondary')
  })

  it('should set height to 100px when height is 100', () => {
    createComponent({
      onClick: () => {},
      height: 100
    })

    expect(wrapper.attributes('style')).toContain('height: 100px')
  })

  it('should set width 100px when width is 100', () => {
    createComponent({
      onClick: () => {},
      width: 100
    })

    expect(wrapper.attributes('style')).toContain('width: 100px')
  })

  it('should set disabled when disabled is true', () => {
    createComponent({
      onClick: () => {},
      disabled: true
    })

    expect(wrapper.attributes()).toHaveProperty('disabled')
  })
})