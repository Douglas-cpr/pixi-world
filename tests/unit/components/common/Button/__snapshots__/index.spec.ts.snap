// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render correctly 1`] = `
<button
  class="common__button common__button--primary"
  style="height: 40px; width: 100px;"
>
  
  
</button>
`;
